# TABLE OF CONTENTS
### 01. Importing Libraries
### 02. Importing Data
### 03. Task

# 01. Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
path = r'C:\Users\gabri\OneDrive\Desktop\CF Data Analysys Course\Data Analytics Immersion\A4 Achivement\Instacart Basket Analysis'

In [3]:
# products dataframe
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# orders dataframe
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'),  index_col = False)

In [5]:
# dataframes in this notebook
    # df_ords
    # df_prods
    # df_test - created for exercise training
    # df_prods_nan - created to see nulls
    # df_prods_clean - without missing values
    # df_dups - created to find duplicates
    # df_prods_clean_no_dups 

In [6]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_sequence_per_user  \
0           0   2539329        1    prior                        1   
1           1   2398795        1    prior                        2   
2           2    473747        1    prior                        3   
3           3   2254736        1    prior                        4   
4           4    431534        1    prior                        5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

In [7]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_sequence_per_user  \
count  3.421083e+06  3.421083e+06  3.421083e+06             3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05             1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04             1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00             1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04             5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05             1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05             2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05             1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [9]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [10]:
# create a dataframe
df_test = pd.DataFrame()

In [11]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# check for mixed types old
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\gabri\AppData\Local\Temp\ipykernel_9564\1290911561.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [14]:
# check for mixed types new
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
# check for mixed types new
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].map(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [16]:
# changing type in column
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
# check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
# create a subset to see missing values
df_prods_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_prods_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [21]:
df_prods.shape

(49693, 5)

In [22]:
# creating a new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
df_prods_clean.shape

(49677, 5)

In [24]:
# creating a new dataframe with the duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
df_prods_clean.shape

(49677, 5)

In [27]:
# creating dataframe without dups (from the already clean one)
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [28]:
df_prods_clean_no_dups.shape

(49672, 5)

In [29]:
# export prepared dataset
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 03. Task

In [30]:
# step 2
df_ords.describe()

Unnamed: 0      order_id       user_id  order_sequence_per_user  \
count  3.421083e+06  3.421083e+06  3.421083e+06             3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05             1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04             1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00             1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04             5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05             1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05             2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05             1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [31]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_sequence_per_user  \
0           0   2539329        1    prior                        1   
1           1   2398795        1    prior                        2   
2           2    473747        1    prior                        3   
3           3   2254736        1    prior                        4   
4           4    431534        1    prior                        5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

## Step 2 answer: days_since_prior max is odd (Why not more than 30days between order?). order_sequence_per_user max is odd (no one ever made more than 100 orders?). Have extra column Unnamed that needs to drop (old index). Eval_set can be dropped (according to old exercises) - done at the end before export

In [32]:
# step 3 - looking for mixed data type (using new formula)
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Step 3 and 4 answer: no mixed data type found

In [33]:
# step 5 - looking for missing values
df_ords.isnull().sum()

Unnamed: 0                      0
order_id                        0
user_id                         0
eval_set                        0
order_sequence_per_user         0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order     206209
dtype: int64

In [34]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_sequence_per_user  \
0           0   2539329        1    prior                        1   
1           1   2398795        1    prior                        2   
2           2    473747        1    prior                        3   
3           3   2254736        1    prior                        4   
4           4    431534        1    prior                        5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

## Step 5 Answer: Only column with missing values is days_since_prior_order (206209 count). Those are the first time orders made by each user_id

## Step 6 Answer: I decided not to remove or change anything on it, since it "makes sense" to be missing, as there were no prior order. I debated with myself about imputting 0, but then it could create confusion with orders made on the same day, as it would be a 0 in that column and could generate false duplicates (first order of the user_id the exact same as a subsequent order made on the same day as another order)

In [35]:
# address the mising values using an appropriate method - given by Halima
df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True
df_ords_clean

Unnamed: 0  order_id  user_id eval_set  order_sequence_per_user  \
0                 0   2539329        1    prior                        1   
1                 1   2398795        1    prior                        2   
2                 2    473747        1    prior                        3   
3                 3   2254736        1    prior                        4   
4                 4    431534        1    prior                        5   
...             ...       ...      ...      ...                      ...   
3421078     3421078   2266710   206209    prior                       10   
3421079     3421079   1854736   206209    prior                       11   
3421080     3421080    626363   206209    prior                       12   
3421081     3421081   2977660   206209    prior                       13   
3421082     3421082    272231   206209    train                       14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         3                  7                    15.0   
2                         3                 12                    21.0   
3                         4                  7                    29.0   
4                         4                 15                    28.0   
...                     ...                ...                     ...   
3421078                   5                 18                    29.0   
3421079                   4                 10                    30.0   
3421080                   1                 12                    18.0   
3421081                   1                 12                     7.0   
3421082                   6                 14                    30.0   

         new_customer  
0                True  
1               False  
2               False  
3               False  
4               False  
...               ...  
3421078         False  
3421079         False  
3421080         False  
3421081         False  
3421082         False  

[3421083 rows x 9 columns]

In [36]:
df_ords_clean['new_customer'].value_counts()

new_customer
False    3214874
True      206209
Name: count, dtype: int64

In [36]:
# step 7 - finding duplicates
df_dups = df_ords[df_ords.duplicated()]

In [37]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_sequence_per_user, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

## Step 7 and 8 Answer: no duplicates found in the dataframe, no action needed

In [38]:
# dropping unnecessary columns before exporting - Unnamed and Eval_set
df_ords_clean = df_ords.drop(columns = ['Unnamed: 0', 'eval_set'])

In [39]:
df_ords_clean

order_id  user_id  order_sequence_per_user  orders_day_of_week  \
0         2539329        1                        1                   2   
1         2398795        1                        2                   3   
2          473747        1                        3                   3   
3         2254736        1                        4                   4   
4          431534        1                        5                   4   
...           ...      ...                      ...                 ...   
3421078   2266710   206209                       10                   5   
3421079   1854736   206209                       11                   4   
3421080    626363   206209                       12                   1   
3421081   2977660   206209                       13                   1   
3421082    272231   206209                       14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

In [40]:
# export prepared dataset
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [41]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [42]:
df_ords_clean

order_id  user_id  order_sequence_per_user  orders_day_of_week  \
0         2539329        1                        1                   2   
1         2398795        1                        2                   3   
2          473747        1                        3                   3   
3         2254736        1                        4                   4   
4          431534        1                        5                   4   
...           ...      ...                      ...                 ...   
3421078   2266710   206209                       10                   5   
3421079   1854736   206209                       11                   4   
3421080    626363   206209                       12                   1   
3421081   2977660   206209                       13                   1   
3421082    272231   206209                       14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]